docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1119 @@
+store knowledge about up to which point the deletions have been validated.
+
+In addition to the lists themselves, a header object would be used
+to store state about how far validation has proceeded (in the "Option C" case
+above, for background validation), and to record the next sequence number in
+case there are no deletion lists at present.
+
+Deletion queue objects will be stored outside the `tenants/` path, and
+with the node generation number in the name. The paths would look something like:
+
+```bash
+  # Deletion List
+  deletion/<node id>/<sequence>-<generation>.list
+
+  # Header object, stores the highest sequence & clean sequence
+  deletion/<node id>/header-<generation>
+```
+
+Each entry in a deletion list is structured to contain the tenant & timeline,
+and the node generation & attachment generation.
+
+```rust
+/// One of these per deletion list object
+struct DeletionList {
+  deletions: Vec<DeletionEntry>
+  node_gen: NodeGeneration
+}
+
+/// N of these per deletion list
+struct DeletionEntry {
+  tenant_id: TenantId,
+  timeline_id: TimelineId,
+  keys: Vec<String>,
+  attach_gen: AttachmentGeneration
+}
+```
+
+#### Deletion queue replay on startup
+
+When starting up with a new generation number, a pageserver should avoid
+leaking objects by ingesting the deletion queue from its previous lifetime, but
+ignore any un-validated content:
+
+1. List all objects in `deletion/<node id>`. Fetch the header.
+2. Drop any that are not below the validated horizon in the header
+3. Write a new header in the current generation number
+4. Process the validated lists from the previous generation in the same
+   way as any lists that would be generated within this generation.
+
+The number of objects leaked in this process depends on how frequently we
+do validation during normal operations, and whether the previous pageserver
+instance was terminated cleanly and validated its lists in the process.
+
+#### Background generation validation
+
+Deletion execution is gated on the "validated sequence" advancing. To advance it,
+some background task would do the following procedure for each deletion list
+in order:
+
+1. Scan the DeletionList and aggregate a map of tenant to attachment generation
+2. Send a request to the control plane to validate these generations and the
+   DeletionList's node generation.
+3. Conditional on result:
+   - a. If all are valid, then advance validated sequence number
+   - b. If some are invalid, then rewrite the deletion list to exclude all non-valid
+     deletions, persist to S3, then advance the validated sequence number.
+
+#### Operations that may skip the queue
+
+Deletions of an entire timeline are exempt from generation number validation. Once the
+control plane sends the deletion request, there is no requirement to retain the readability
+of any data within the timeline, and all objects within the timeline path may be deleted
+at any time from the control plane's deletion request onwards.
```

Comment:
What if the pageserver that was first requested to do the deletions dies mid-way through the deletions?
Do we continue to rely on existing `deleted_at` mechanism and carefulness about deleting index part last, so that we don't leak objects?
Worth pointing out, I think.

Also, in terms of visibility: what if there's an earlier unresponsive attachment that still has the timeline visible that the current attachment is deleting? That earlier attachment may write again even after the current attachment did a one-off delete-entire-prefix. I guess we rely on the scrubber?

Finally: what index parts does an attachment delete? just its own one, or does it opportunistically delete all of them?

Reply:
> Finally: what index parts does an attachment delete? just its own one, or does it opportunistically delete all of them?

Ah, I guess that's covered by `Cleaning up previous generations' remote indices`.

Reply:
> What if the pageserver that was first requested to do the deletions dies mid-way through the deletions?
Do we continue to rely on existing deleted_at mechanism and carefulness about deleting index part last, so that we don't leak objects?

This section doesn't change the existing behavior for timeline deletions -- we're basically just saying that `delete_all` gets to skip all the validation.

> Also, in terms of visibility: what if there's an earlier unresponsive attachment that still has the timeline visible that the current attachment is deleting? That earlier attachment may write again even after the current attachment did a one-off delete-entire-prefix. I guess we rely on the scrubber?

Right, that's a scrubber case.

> Finally: what index parts does an attachment delete? just its own one, or does it opportunistically delete all of them?

This gets handled by the existing `delete_all` behavior where it lists objects under a timeline and deletes whatever it finds, as its last step in remote deletion.